api/server/middleware/middleware.go

```diff
@@ -1,7 +1,13 @@
 package middleware
 
-import "github.com/docker/docker/api/server/httputils"
+import (
+	"net/http"
 
-// Middleware is an adapter to allow the use of ordinary functions as Docker API filters.
-// Any function that has the appropriate signature can be registered as a middleware.
-type Middleware func(handler httputils.APIFunc) httputils.APIFunc
+	"golang.org/x/net/context"
+)
+
+// Middleware is an interface to allow the use of ordinary functions as Docker API filters.
+// Any struct that has the appropriate signature can be registered as a middleware.
+type Middleware interface {
+	WrapHandler(func(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error) func(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error
```

Comment:
I misspoke above. I meant "To avoid references *from* `pkg`'.

There is a build script [`validate-pkg`](https://github.com/docker/docker/blob/master/hack/make/validate-pkg) that exists to prevent references from `pkg` to internal docker stuff. The thought is that externals `pkg` should stand alone.

I might be over thinking this. No :coffee: this morning.